internal/terraform/workspace/workspace.go

```diff
@@ -0,0 +1,284 @@
+	}
+
+	p, err := lang.FindCompatibleParser(version)
+	if err != nil {
+		return err
+	}
+	p.SetLogger(w.logger)
+
+	ss := w.newSchemaStorage()
+
+	ss.SetLogger(w.logger)
+
+	p.SetSchemaReader(ss)
+
+	w.parser = p
+	w.schemaWriter = ss
+	w.tfExec = tf
+	w.tfVersion = version
+
+	err = w.initPluginCache(dir)
+	if err != nil {
+		return fmt.Errorf("plugin initialization failed: %w", err)
+	}
+	err = w.initModuleCache(dir)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (w *workspace) initTfExecutor(dir string) (*exec.Executor, error) {
+	tfPath := w.tfExecPath
+	if tfPath == "" {
+		var err error
+		tfPath, err = w.tfDiscoFunc()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	tf := w.tfNewExecutor(w.ctx, tfPath)
+
+	tf.SetWorkdir(dir)
+	tf.SetLogger(w.logger)
+
+	if w.tfExecLogPath != "" {
+		tf.SetExecLogPath(w.tfExecLogPath)
+	}
+
+	if w.tfExecTimeout != 0 {
+		tf.SetTimeout(w.tfExecTimeout)
+	}
+
+	return tf, nil
+}
+
+func (w *workspace) initPluginCache(dir string) error {
+	var lf *TrackedFile
+	if w.ignorePluginCache {
+		lf = &TrackedFile{
+			Path: pluginLockFilePaths(dir)[0],
+		}
+	} else {
+		var err error
+		lockPaths := pluginLockFilePaths(dir)
+		lf, err = findTrackedFile(lockPaths)
+		if err != nil {
+			if os.IsNotExist(err) {
+				return &tferr.NotInitializedErr{Dir: dir}
+			}
+
+			return fmt.Errorf("unable to calculate hash: %w", err)
+		}
+	}
+
+	return w.UpdatePluginCache(lf)
+}
+
+func (w *workspace) initModuleCache(dir string) error {
+	lf, err := trackedFileFromPath(moduleLockFilePath(dir))
+	if err != nil {
+		if os.IsNotExist(err) {
+			w.logger.Printf("module lock file not found: %s", err.Error())
+			return nil
+		}
+
+		return fmt.Errorf("unable to calculate hash: %w", err)
+	}
+
+	return w.UpdateModuleCache(lf)
+}
+
+func (w *workspace) UpdateModuleCache(lockFile *TrackedFile) error {
+	w.moduleMu.Lock()
+	w.logger.Printf("updating module cache based on %s ...", lockFile.Path)
+	defer w.moduleMu.Unlock()
+
+	w.moduleLockFile = lockFile
+
+	b, err := ioutil.ReadFile(lockFile.Path)
+	if err != nil {
+		return err
+	}
+	mc := ModuleCache{}
+	err = json.Unmarshal(b, &mc)
+	if err != nil {
+		return err
+	}
+	mc.rootDir = workspaceDirFromFilePath(lockFile.Path)
+	w.moduleCache = &mc
+	w.logger.Printf("updated module cache - %d references parsed", len(mc.Modules))
+	return nil
+}
+
+func (w *workspace) Parser() lang.Parser {
+	w.pluginMu.RLock()
+	defer w.pluginMu.RUnlock()
+
+	return w.parser
+}
+
+func (w *workspace) ReferencesModulePath(path string) bool {
+	if w.moduleCache == nil {
+		return false
+	}
+
+	for _, m := range w.moduleCache.Modules {
+		if m.Key == "" {
+			// skip root module, as that's tracked separately
+			continue
+		}
+		if !strings.HasPrefix(m.Source, "."+string(os.PathSeparator)) {
+			// skip external modules, for now
+			continue
+		}
+		absPath := filepath.Join(w.moduleCache.rootDir, m.Dir)
+		w.logger.Printf("checking if %q == %q", absPath, path)
+		if absPath == path {
```

Comment:
If you are going to do a string comparison here, you probably want to use `filepath.Clean`. Not sure if this still leaves an edge for symlinks though.

Reply:
True, I think the right side (`path`) can be cleaned up in the manager's `WorkspaceByPath`, but no harm in cleaning it up here again I guess.

The left part certainly needs cleaning here though. I need to better understand what paths we may expect here first. I'm not sure if Terraform itself does any cleaning and/or symlink eval before storing these paths in the cache.

Reply:
`w.moduleCache.rootDir` is cleaned as part of `TrackedFile` constructor, `m.Dir` as part of manifest parser (although Terraform should produce clean paths in general) and the `path` is cleaned in `WorkspaceByPath` and `AddWorkspace`.